### DISCUSSION TITLE
antlr grammar license?

### DISCUSSION THREAD

**Discussion by xied75** [dotnet/csharplang#773] [General]
Dear team,

Just want to clarify the licensing model on the antlr grammar in the specs markdown? e.g. for the C# 6 draft  spec. Would it be ok for the antlr community to use those grammars in a suitable open license?

Currently there is a set of g4 files here: https://github.com/antlr/grammars-v4
https://github.com/antlr/grammars-v4/tree/master/csharp

It would be great if the work can be unified into One.

p.s. By any chance all those segments we saw over multiple markdowns are in fact in a single file form somewhere?

**Comment by svick** [comment]
The spec was originally converted to Markdown and grammar to ANTLR in [ljw1004/csharpspec](https://github.com/ljw1004/csharpspec), which also contains [the grammar in a single file](https://github.com/ljw1004/csharpspec/blob/gh-pages/csharp.g4).

And I agree it would be great if the spec and grammar were licensed under an open license, but I think [this file](https://github.com/dotnet/csharplang/blob/master/spec/LICENSE.md) makes it clear that's not currently the case.

**Comment by 4creators** [comment]
How does this license statement relates to the fact that C# language specification is an approved international standard? Is this an overreaching copyright declaration? To what extent such copyright claim is invalidated by the fact that specification copyright lies with (i) ECMA and ISO or (ii) in reality specification is in public domain (I have not checked specific wording of copy on standards yet)?

**Comment by svick** [comment]
@4creators I don't think that changes anything. Especially considering that, AFAIK, standards often aren't licensed under open licenses, and certainly aren't released into the public domain. For example, the ISO C++ standard is not even gratis (free as in beer), let alone libre (free as in speech).

Specifically, I couldn't find any kind of license for [the ECMA version of the C# 2.0 spec](http://www.ecma-international.org/flat/publications/Standards/Ecma-334.htm) and [the ISO version of the C# 2.0 spec has a really draconian license](http://standards.iso.org/ittf/PubliclyAvailableStandards/c042926_ISO_IEC_23270_2006(E).zip):

> The document is a single-user, non-revisable Adobe Acrobat® PDF file. You are downloading a single-user licence to store this file on your personal computer. Under no circumstances may the electronic file you are licensing be copied, transferred, or placed on a network of any sort without the authorization of the copyright owner.
>
> You may print out and retain one-only printed copy of the PDF file.

**Comment by jnm2** [comment]
> The document is a single-user, non-revisable Adobe Acrobat® PDF file. You are downloading a single-user licence to store this file on your personal computer. Under no circumstances may the electronic file you are licensing be copied, transferred, or placed on a network of any sort without the authorization of the copyright owner.
> 
> You may print out and retain one-only printed copy of the PDF file.

Who comes up with this stuff? 😆

**Comment by iam3yal** [comment]
@jnm2 lawyers, ofc! 😆

**Comment by 4creators** [comment]
> The unauthorized reproduction or distribution of this copyrighted work is illegal and may be punishable by criminal law.

That is another part of ISO licence language, however, I doubt it is enforceable with regard to syntax which is in the document as any reproduction and use i.e. for lexer/parser code generation is an intended use of the standard itself.  But it is even more interesting!

This same reasoning may be applied to ECMA standard but ECMA even did not not claim any copyright on published standard and it was published before ISO standard. Therefore anything what is there is automatically in public domain and consequently intersection of ISO and ECMA is in public domain too. Honestly I do not care what ISO claims with regard to C# standard as such claims would be thrown away as frivolous by any court in any country including Switzerland. 

Below is the disclaimer from Microsoft published annotated ECMA standard:

> You may download without any restriction annotated by Microsoft ECMA standards. Inside there is an statement:This fourth edition cancels and replaces the third edition. Changes from the previous edition were made to align this Standard with ISO/IEC 23271:2006. The following companies and organizations have participated in the development of this standard, and their contributions are gratefully acknowledged: Borland, Fujitsu Software Corporation, Hewlett-Packard, Intel Corporation, IBM Corporation, ISE, IT University of Copenhagen, Jagger Software Ltd., Microsoft Corporation, Monash University, Netscape, Novell/Ximian, Phone.Com, Plum Hall, Sun Microsystems, and University of Canterbury (NZ)

Lack of copyright on any of documents produced by ECMA and Microsoft and statement as below:

> You may download without any restriction annotated by Microsoft ECMA standards

moves it into public domain. It is the same with photography -> publish picture without copyright notice and it is in public domain automatically.

The directory where Microsoft placed copy notice contains derivative work based on work which is in public domain and AFAIR you cannot take anything out of public domain once it is there.

Very interesting legal case regarding attempt to copyright public domain artifacts is a [Getty Likely To Settle $1B Suit By Photographer For Appropriating Her Public-Domain Work](https://www.forbes.com/sites/legalentertainment/2016/08/03/pay-up-getty-sends-trolling-letter-to-photographer-highsmith-demanding-money-for-her-own-photos/#2ce36afd596e).

**Comment by svick** [comment]
@4creators 

> Lack of copyright on any of documents produced by ECMA and Microsoft […]
> moves it into public domain. It is the same with photography -> publish picture without copyright notice and it is in public domain automatically.

That's really not how copyright works. [Since 1989, copyright notices are optional in the US.](https://en.wikipedia.org/wiki/Copyright_notice)

**Comment by 4creators** [comment]
But US is not a whole of the world and if you distribute work via internet from Switzerland ... ECMA is located in Switzerland and [the relations between Switzerland and the European Union (EU) are framed by a series of bilateral treaties whereby the Swiss Confederation has adopted various provisions of European Union law in order to participate in the Union's single market](https://en.wikipedia.org/wiki/Switzerland%E2%80%93European_Union_relations). In EU there is no optional copyright notice law, there is no DMCA and no software patents and there is strong personal data protection and net neutrality. ECMA standards are therefore not covered by major parts of US laws and this may create some confusion for users used to work based on US legislation.

In my opinion it's very unfortunate that we have to deal with that sort of problems but right now it is the reality with little hope for any improvement ...

**Comment by svick** [comment]
@4creators 

> But US is not a whole of the world and if you distribute work via internet

Yes, but [most of the world adopted the Berne convention](https://en.wikipedia.org/wiki/Copyright_formalities), which means copyright notices are not necessary.

> at least in EU there is huge distinction between work having copy notice or not having one

I don't believe there is. Do you have some source backing that claim?

**Comment by 4creators** [comment]
No I dont have sources to back this claim and cannot provide them in reasonable time bcs I am not a lawyer :) It is just reasoning about available facts
